### PR TITLE
Top up voucher

### PR DESCRIPTION
**Pull request by jeremyjams**

### CODE REVIEW

File: contracts/VoucherHub.sol [hunk 1]
```diff
@@ -177,6 +177,25 @@ contract VoucherHub is
         emit VoucherCreated(voucherAddress, owner, voucherExpiration, voucherType, value);
     }
 
+    /**
+     * Top up a voucher by increasing its balance and extending the validity
+     * period before expiration.
```
Comment (zguesmi) [inline contracts/VoucherHub.sol:182]: Minor & optionl (just trying to make it more straightforward).
```suggestion
     * Top up a voucher by increasing its balance and pushing its expiration date.
```

File: contracts/VoucherHub.sol [hunk 2]
```diff
@@ -177,6 +177,25 @@ contract VoucherHub is
...
+    function topUpVoucher(address voucher, uint256 value) external onlyRole(VOUCHER_MANAGER_ROLE) {
+        VoucherHubStorage storage $ = _getVoucherHubStorage();
+        require($._isVoucher[voucher], "VoucherHub: unknown voucher");
+        if (value > 0) {
+            _mint(voucher, value); // VCHR
+            IERC20($._iexecPoco).transfer(voucher, value); // SRLC
+        }
+        uint256 voucherExpiration = block.timestamp +
+            $.voucherTypes[Voucher(voucher).getType()].duration;
+        Voucher(voucher).setExpiration(voucherExpiration);
```
Comment (zguesmi) [inline contracts/VoucherHub.sol:195]: Is it cheaper to use one instance of `Voucher(voucher)`?

File: contracts/IVoucherHub.sol [hunk 3]
```diff
@@ -15,6 +15,7 @@ interface IVoucherHub {
         uint256 voucherType,
         uint256 value
     );
+    event VoucherToppedUp(address indexed voucher, uint256 value, uint256 voucherExpiration);
```
Comment (zguesmi) [inline contracts/IVoucherHub.sol:18]: Can we use the same order in `VoucherCreated` and `VoucherToppedUp`?
Update the 1st to be
```solidity
VoucherCreated(voucherAddress, owner, voucherType, voucherExpiration, value)
```
and VoucherToppedUp to
```solidity
VoucherToppedUp(voucher, voucherExpiration, value)
```

File: contracts/VoucherHub.sol [hunk 4]
```diff
@@ -177,6 +177,25 @@ contract VoucherHub is
         emit VoucherCreated(voucherAddress, owner, voucherExpiration, voucherType, value);
     }
 
+    /**
+     * Top up a voucher by increasing its balance and extending the validity
+     * period before expiration.
+     * @param voucher The address of the voucher.
+     * @param value The amount of credits to top up.
+     */
+    function topUpVoucher(address voucher, uint256 value) external onlyRole(VOUCHER_MANAGER_ROLE) {
```
Comment (zguesmi) [inline contracts/VoucherHub.sol:186]: A small suggestion to make the role name more meaningful
(Can be in a separate PR if not in-scope).
```suggestion
    function topUpVoucher(address voucher, uint256 value) external onlyRole(VOUCHERS_MANAGER_ROLE) {
```

File: contracts/beacon/IVoucher.sol [hunk 5]
```diff
@@ -32,6 +33,7 @@ interface IVoucher {
     function getVoucherHub() external view returns (address);
     function getType() external view returns (uint256);
     function getExpiration() external view returns (uint256);
+    function setExpiration(uint256 expiration) external;
```
Comment (zguesmi) [inline contracts/beacon/IVoucher.sol:36]: To put with non-view functions (same in implementation).
We can also decide to not follow the standard layout, we just need to agree on it.

File: test/VoucherHub.test.ts [hunk 6]
```diff
@@ -659,6 +664,86 @@ describe('VoucherHub', function () {
...
+                .withArgs(voucherAddress, topUpValue, expectedExpiration);
+            const voucherCreditBalanceAfter = await voucherHub.balanceOf(voucherAddress);
+            const voucherRlcBalanceAfter = await iexecPocoInstance.balanceOf(voucherAddress);
+            expect(voucherCreditBalanceAfter)
+                .equal(voucherCreditBalanceBefore + topUpValue)
+                .equal(voucherRlcBalanceBefore + topUpValue)
+                .equal(voucherRlcBalanceAfter);
+            expect(
+                await Voucher__factory.connect(voucherAddress, anyone).getExpiration(),
+            ).to.be.equal(expectedExpiration);
```
Comment (zguesmi) [inline test/VoucherHub.test.ts:705]: Also compare to "before" expiration.

File: contracts/VoucherHub.sol [hunk 7]
```diff
@@ -177,6 +177,25 @@ contract VoucherHub is
...
+        VoucherHubStorage storage $ = _getVoucherHubStorage();
+        require($._isVoucher[voucher], "VoucherHub: unknown voucher");
```
Comment (gfournieriExec) [inline contracts/VoucherHub.sol:188]: is it possible to merge it with the modifier onlyVoucher()

File: contracts/VoucherHub.sol [hunk 8]
```diff
@@ -177,6 +177,25 @@ contract VoucherHub is
...
+        if (value > 0) {
```
Comment (gfournieriExec) [inline contracts/VoucherHub.sol:189]: is there a case where value is not > 0 ?

File: test/beacon/Voucher.test.ts [hunk 9]
```diff
@@ -209,6 +210,41 @@ describe('Voucher', function () {
...
+                await voucherAsAnyone.connect(voucherHubSigner).setExpiration(expirationAfter),
+            )
+                .to.emit(voucherAsAnyone, 'ExpirationUpdated')
+                .withArgs(expirationAfter);
+            expect(await voucherAsAnyone.getExpiration())
+                .equal(expirationAfter)
+                .not.equal(expirationBefore);
+        });
+
+        it('Should not set expiration', async function () {
```
Comment (zguesmi) [inline test/beacon/Voucher.test.ts:241]:
```suggestion
        it('Should not set expiration when sender is not authorized', async function () {
```

File: contracts/VoucherHub.sol (hunk 7 above)
Comment (jeremyjams) [reply contracts/VoucherHub.sol:188]: No it is not, `onlyVoucher` is dedicated to `msg.sender` filtering

File: contracts/VoucherHub.sol (hunk 8 above)
Comment (jeremyjams) [reply contracts/VoucherHub.sol:189]: I expect request where user wants someone to extend its credits without recharging value
Do you (& @zguesmi too) want me to remove the `if (value > 0)` so worst case (from gas point of view) is triggering transfer with zero value?
Other point, Imo it might not be a good idea to forbid empty value, it looks best to let it open

File: contracts/VoucherHub.sol (hunk 2 above)
Comment (jeremyjams) [reply contracts/VoucherHub.sol]: Exactly same price (min, max & avg)!
Do you want a syntax update here?

File: contracts/beacon/IVoucher.sol [hunk 10]
```diff
@@ -6,6 +6,7 @@ import {IexecLibOrders_v5} from "@iexec/poco/contracts/libs/IexecLibOrders_v5.so
 pragma solidity ^0.8.20;
 
 interface IVoucher {
+    event ExpirationUpdated(uint256 expiration);
```
[the remark this answers is not in the document]
Comment (jeremyjams) [reply contracts/beacon/IVoucher.sol:9]: I try to sort it related to storage order here

File: contracts/IVoucherHub.sol (hunk 3 above)
Comment (jeremyjams) [reply contracts/IVoucherHub.sol]: Agree! See:
- [Update VoucherCreated event args order](https://github.com/iExecBlockchainComputing/iexec-voucher-contracts/pull/23/commits/b1d8ac9fdd2b0c03e2501dd8feb95f28e3c3249a)
- [Update VoucherToppedUp event args order and use short expiration wording since not ambiguous](https://github.com/iExecBlockchainComputing/iexec-voucher-contracts/pull/23/commits/cad9bb79fbc5931ede64d3ce5e9859a3b8013f7e)

File: contracts/VoucherHub.sol (hunk 4 above)
Comment (jeremyjams) [reply contracts/VoucherHub.sol:186]: See [Add todo to rename roles](https://github.com/iExecBlockchainComputing/iexec-voucher-contracts/pull/23/commits/2a068f388ae2127dcd471eba7506fa55fe14c7cd) :)

File: contracts/beacon/IVoucher.sol (hunk 5 above)
Comment (jeremyjams) [reply contracts/beacon/IVoucher.sol]: See [Move up setExpiration](https://github.com/iExecBlockchainComputing/iexec-voucher-contracts/pull/23/commits/010660fb95995c735e160cf607214326b31c4e16)

File: contracts/beacon/Voucher.sol [hunk 11]
```diff
@@ -255,6 +261,16 @@ contract Voucher is Initializable, IVoucher {
         return $._expiration;
     }
 
+    /**
+     * Set the expiration timestamp of the voucher.
+     * @param expiration The expiration timestamp.
+     */
+    function setExpiration(uint256 expiration) external onlyVoucherHub {
```
[the remark this answers is not in the document]
Comment (jeremyjams) [reply contracts/beacon/Voucher.sol]: Yes, see [Move up setExpiration](https://github.com/iExecBlockchainComputing/iexec-voucher-contracts/pull/23/commits/010660fb95995c735e160cf607214326b31c4e16)!

File: test/VoucherHub.test.ts (hunk 6 above)
Comment (jeremyjams) [reply test/VoucherHub.test.ts]: Please see [Compare to "before" expiration](https://github.com/iExecBlockchainComputing/iexec-voucher-contracts/pull/23/commits/a9f3362360a05b1d8bc6a6256280b97caa69ae28) :)

File: contracts/VoucherHub.sol (hunk 2 above)
Comment (zguesmi) [reply contracts/VoucherHub.sol]: No it's ok like that. Thank you testing it.

File: contracts/VoucherHub.sol (hunk 8 above)
Comment (zguesmi) [reply contracts/VoucherHub.sol:189]: How about splitting into:
1. topUpVoucher(voucher, value) => fails if not value == 0
2. extandExpiration(voucher)

**Review by gfournieriExec — APPROVED** [review]
Thanks for the feature !

**Review by zguesmi — APPROVED** [review]
Nice!
